libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLTreeDigest.java

```diff
@@ -0,0 +1,441 @@
+                    bWeight = b.count();
+
+                    assert bMean >= aMean;
+                } else {
+                    bWeight = 0;
+                }
+            }
+            // shouldn't be possible because x <= lastMean
+            throw new IllegalStateException("Ran out of centroids");
+        }
+    }
+
+    private double interpolateTail(AVLGroupTree values, double x, int node, double mean, double extremeValue) {
+        int count = values.count(node);
+        assert count > 1;
+        if (count == 2) {
+            // other sample must be on the other side of the mean
+            return 1.0 / size();
+        } else {
+            // how much weight is available for interpolation?
+            double weight = count / 2.0 - 1;
+            // how much is between min and here?
+            double partialWeight = (extremeValue - x) / (extremeValue - mean) * weight;
+            // account for sample at min along with interpolated weight
+            return (partialWeight + 1.0) / size();
+        }
+    }
+
+    /**
+     * @param q The quantile desired.  Can be in the range [0,1].
+     * @return The minimum value x such that we think that the proportion of samples is &le; x is q.
+     */
+    @Override
+    public double quantile(double q) {
+        if (q < 0 || q > 1) {
+            throw new IllegalArgumentException("q should be in [0,1], got " + q);
+        }
+
+        AVLGroupTree values = summary;
+        if (values.size() == 0) {
+            // no centroids means no data, no way to get a quantile
+            return Double.NaN;
+        } else if (values.size() == 1) {
+            // with one data point, all quantiles lead to Rome
+            return values.iterator().next().mean();
+        }
+
+        // if values were stored in a sorted array, index would be the offset we are interested in
+        final double index = q * (count - 1);
+
+        // deal with min and max as a special case singletons
+        if (index <= 0) {
+            return min;
+        }
+
+        if (index >= count - 1) {
+            return max;
+        }
+
+        int currentNode = values.first();
+        int currentWeight = values.count(currentNode);
+
+        // weightSoFar represents the total mass to the left of the center of the current node.
+        // If there's a singleton on the left boundary, there's no more mass on the left.
+        double weightSoFar = (currentWeight == 1) ? 0 : currentWeight / 2.0;
+
+        if (index <= weightSoFar || (currentWeight == 1 && index <= currentWeight)) {
+            // If there's a singleton on the left boundary, interpolate between the first two centroids.
+            if (currentWeight == 1) {
+                return min + index * (values.mean(values.next(currentNode)) - min);
+            }
+            // Interpolate between min and first mean, otherwise.
+            return (min * index + values.mean(currentNode) * (weightSoFar - index)) / weightSoFar;
```

Comment:
should we use the `weightedAverage` helper method here too for consistency?

Reply:
Totally, updated here and in `MergingDigest`.